### ISSUE TITLE
Provide a dummy ```jax.key(...)``` that allows splitting, but not sampling

### ISSUE THREAD

**Issue by knyazer** [google/jax#23445]
Hey guys,

When implementing a method that would statically (python control-flow) do something with a key, for example Dropout, it seems to be much neater to use some sort of dummy key, that would allow splitting, but don't allow sampling from it (meaning, statically throw). Here is an example:
```py
@jax.jit(static_argnums=(1,))
def dropout(x, p, key):
  if p == 0.0:
    return x
  else:
    assert key is not None # get a decent error if no key given, instead of a weird error down the stack
    return actual_dropout(x, p, key)
y = jnp.zeros((5,))
x = dropout(y, 0.1, jr.key(0)) # all is good, ergonomic, etc etc
x = dropout(y, 0.0, None) # sad and ugly
x = dropout(y, 0.1, None) # should throw a meaningful error

# so far so good, but if we try to vmap...
v_drop = jax.vmap(dropout, static_argnums=(1,))
key = jr.key()
x = v_drop([y, y], [0.1, 0.2], jr.split(key, 2)) # fine
x = v_drop([y, y], [0.0, 0.0], [None, None]) # obviously does not work, and ugly
```
What I usually do is a use a placeholder key (eg jr.key(42)) but that _possibly_ allows sampling from it, and i don't want that. The 'correct' way is to use Nones, but they force you to use too many switches in logic (e.g too many conditions in the final code, does not allow you to easily switch between the two modes), so my proposal is to have jr.dummy(0) that is splittable, but if one tries to sample it -> it throws tracing time. This way we can adapt the above example naively with:
```py
key = jr.dummy(0) # or jr.key(0)
x = v_drop([y, y], [0.1, 0.2], jr.split(key, 2)) # fine
x = v_drop([y, y], [0.0, 0.0], jr.split(key, 2)) # also fine 
```

I am happy to hear any critique, but it seems this should be pretty easy to implement: just attach some static property to a key when ```dummy``` and then when sampling check for this dummy property.

**Comment by jakevdp** [comment]
Hi @knyazer – I'm having a bit of trouble following the logic in your code: for example, you appear to be passing lists to `vmap`, in which case the function will map in parallel over the contents of each list item, not map over the entries in the list. Also, you use `static_argnums` within your `vmap` call, which is not a valid parameter.

I suspect this is meant to be pseudocode illustrative of your intent, but I think your intent is obscured by its invalidity: it's not clear to me what your actual problem is.

Perhaps you could edit your code to something that is executable to better show your intent?

**Comment by knyazer** [comment]
Sorry, Jake! The reason I used an invalid-pseudo-jax code is because these days I mostly use Equinox, and I totally don't remember the syntax of the pure JAX :) Anyhow...

After dreaming up a "working" minimal example, I failed to actually make it meaningful: my example forces the user to write a singular additional condition, which is certainly not a good reason to introduce new functionality. I will think about it a bit, but, probably, the reason for this feature request is mainly my poor coding skills :)

Closing for now, will reopen if I come up with a good example. Sorry for the noise.
